### ISSUE TITLE
Failed to allocate read-token

### ISSUE THREAD

**Issue by salkhwlani** [cachewerk/relay#153]
## Description

Sometimes, for the same key in the same server, an error returns "Relay\Relay::get(): Failed to allocate read-token, skipping in-memory cache lookup for command."

The error despair once the server restart or the php-fpm process restart

## Expected Behavior

In case the relay fails to allocate the read token, the client should try to look up the command in the remote Redis.

## Actual Behavior

The client threw an exception and repeated for all requests served from the same process until the php-fpm worker restart

## Steps to Reproduce

It's happens randomly for some process

## Relay diagnostics

```
relay

Relay Support => enabled
Relay Version => 0.8.1
Available cache => 1073687488
Available serializers => php, json, igbinary, msgpack
Available compression => lzf, zstd, lz4
Binary UUID => xxxxxxxxxxxxxx
Git SHA =>
Allocator => relay-mmap
Build OS => Alpine Linux v3.17
Architecture => x86_64
OpenSSL version => OpenSSL 3.0.12 24 Oct 2023
relay.enabled => true
relay.environment => production
relay.key => XXXX-XXXX-XXXXXX-XXXXXXX-XXXXXXX-XXXXXXX
relay.maxmemory => 1073741824
relay.maxmemory_pct => 95
relay.eviction_policy => lru
relay.eviction_sample_keys => 128
relay.initial_readers => 128
relay.invalidation_poll_usec => 60000
relay.pconnect_default => 1
relay.max_endpoint_dbs => 60
relay.loglevel => debug
relay.logfile => /var/log/relay.log
relay.session.compression => none
relay.session.compression_level => 0
relay.session.locking_enabled => 0
relay.session.lock_expire => 0
relay.session.lock_wait_time => 0
```

## Environment

* OS and version: Alpine
* PHP version: 8.3.13
* Redis version: 6.1.0

**Comment by tillkruss** [comment]
We fixed a lot of issues in 0.9.0, can you reproduce the issue on that version?

**Comment by salkhwlani** [comment]
@tillkruss I'll update the client to 0.9.0 and see if the error triggered at that version

**Comment by salkhwlani** [comment]
@tillkruss, the issue is still present in client 0.9.0

**Comment by michael-grunder** [comment]
This must be a resource and/or memory leak of some kind.

The "read tokens" in question are how concurrent lock-free reads can be done. We allocate a configurable number of these up front with this `ini` directive:

```ini
# 128 is the default
relay.initial_readers = 128
```

This means that there should always be 128 available to every php worker process.

If there are more than 128 concurrent requests, Relay could emit this message due to being OOM. However, if there are fewer workers we must be leaking them somehow.

Can you provide more information about your webserver setup? Is it configured to cycle/kill the worker processes after a certain number of requests or amount of time?

Perhaps we aren't hooking that event properly and leaking them that way.

**Comment by salkhwlani** [comment]
@michael-grunder, you are correct. The PHP process will restart after processing 3k requests or when it reaches the maximum memory limit, whichever comes first.

btw, when an error occurs, it is repeated in the same process and for the same cache key (it's only happened very rare bases in the overall process are running)

- Mode: dynamic
- Max: 30 process
- Redis Connection per process (Max): 3 endpoint

I believe, based on that, the 128 readers are more than enough, right?

**Comment by michael-grunder** [comment]
I was able to replicate this behavior locally. The fix should be fairly straightforward.

**Comment by michael-grunder** [comment]
Quick update. I've implemented a fix and am fuzzing the implementation now. I should have it merged soon. Then we'll tag a version and do a release.

**Comment by tillkruss** [comment]
Fixed in v0.9.2-dev
